torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -593,51 +591,34 @@ def moe_forward(self, x, topk_ids, topk_weight):
         # TODO: don't use `received`
         gathered_tokens = self.token_gather_buf[:received]
 
-        # This part prepares the token indices for each expert, because we
-        # receive them in a whole chunk for the entire expert group (i.e.,
-        # device). This part doesn't need gradient.
+        # This part prepares a 1D tensor with the same length as
+        # `gathered_tokens`. The 1D tensor is filled with local expert IDs which
+        # the tokens in `gathered_tokens` are headed for. This part doesn't need
+        # gradient.
         with torch.no_grad():
-            tokens_per_expert_post_gather = tokens_per_expert_group.view(
-                self.ep_size, self.experts_per_rank
-            ).sum(dim=0)
             gatherd_idxs = np.zeros(shape=(gathered_tokens.shape[0],), dtype=np.int32)
             s = 0
             # TODO: remove `tolist()`
             for i, k in enumerate(tokens_per_expert_group.tolist()):
                 gatherd_idxs[s : s + k] = i % self.experts_per_rank
                 s += k
-            gatherd_idxs = gatherd_idxs.argsort()
-            # TODO: remove `tolist()`
-            tokens_per_expert = tokens_per_expert_post_gather.tolist()
-
-        sorted_tokens = gathered_tokens[gatherd_idxs]
-
-        outputs = []
-        start_idx = 0
-        for i, num_tokens in enumerate(tokens_per_expert):
-            end_idx = start_idx + num_tokens
-            if num_tokens == 0:
-                continue
-            expert = self.experts[i + self.ep_rank * self.experts_per_rank]
-            tokens_for_this_expert = sorted_tokens[start_idx:end_idx]
-            expert_out = expert(tokens_for_this_expert)
-            outputs.append(expert_out)
-            start_idx = end_idx
-
-        # len(outputs) == 0 means no tokens routed to this EP rank.
-        # `sorted_tokens` would have shape [0, hidden_dim], we use it so that
-        # `outs` is an empty tensor with shape [0, hidden_dim]
-        outs = torch.cat(outputs, dim=0) if len(outputs) else sorted_tokens
-        # Take necessary space from `token_gather_buf` symm mem
-        new_x = self.token_gather_buf[: outs.shape[0]]
-        new_x[gatherd_idxs] = outs
-        gathered_tokens = new_x.new_empty(*sorted_tokens_shape)
+
+        # Take necessary space from `token_send_buf` symm mem because we are
+        # going to send them out after expert processing
+        processed_tokens = self.token_send_buf[:received]
+        for i, expert in enumerate(self.experts.values()):
+            processed_tokens[gatherd_idxs == i] = expert(
+                gathered_tokens[gatherd_idxs == i]
+            )
+
+        # Take necessary space from `token_gather_buf` symm mem to receive processed tokens
+        gathered_tokens = self.token_gather_buf[: sorted_tokens_shape[0]]
```

Comment:
iiuc the token_gather_buf is currently one-per-layer but later PR makes it shared?  regarding 'take necessary space', it does not actually do any bookkeeping bc it assumes there is nobody else using the space, its just grabbing a slice out of the preallocated buffer?

Reply:
Yes, because layers run sequentially.